### PR TITLE
[inductor] Preserve metadata across replace_by_example and register_replacement patterns

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138089

replace_by_example is used to implement some pattern-matching passes in inductor. Previously, replace_by_example would generate nodes with very little metadata. In particular, `meta["original_aten"]` would be lost; that meant that when generating triton kernel names, you could get empty names like `triton_tem_fused_0` if the input nodes to the fused kernel were the result of a pattern-matching pass that used replace_by_example.

This also adds metadata for to register_replacement patterns, including pad_mm.

This fixes the issue by copying metadata from the original node to the replacement nodes. If there are multiple original nodes we skip the metadata transfer; so if you have a `add(z, mm(x, y))`, then the metadata won't be transferred right now.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D64480755](https://our.internmc.facebook.com/intern/diff/D64480755)

### CODE REVIEW

File: torch/_inductor/pattern_matcher.py [hunk 1]
```diff
@@ -139,6 +139,15 @@ def __init__(self) -> None:
 MULTIPLE = Multiple()
 
 
+def _transfer_meta(new_meta: Dict[str, Any], old_meta: Dict[str, Any]) -> None:
+    # transfer metadata after pattern matching occurs.
+    # skip "val" and "tensor_meta" because this info is too specific; it's unlikely
+    # to remain accurate after pattern matching has occurred.
+    new_meta.update(
+        (k, v) for k, v in old_meta.items() if k not in {"val", "tensor_meta"}
+    )
```
Comment (davidberard98) [inline torch/_inductor/pattern_matcher.py:148]: Is there a pre-existing utility to do something like this? And if not - am I missing anything other than "val" and "tensor_meta" that should be skipped?

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64480755).

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (Chillee) [reply torch/_inductor/pattern_matcher.py:148]: We do have https://github.com/pytorch/pytorch/blob/9b01d17b8d7771bb0501c1389759d726b5533557/torch/fx/proxy.py#L93

although I'm not sure it's directly applicable.

The other thing is that, in some places, we do use the FakeTensorUpdater to ensure that the faketensor metadata is accurate.

Comment (davidberard98) [reply torch/_inductor/pattern_matcher.py:148]: great, thanks! `_COPY_META_FIELDS` seems good enough to me. It looks like the faketensor metadata propagation is already handled in these cases and I don't see any issues with it, so I'll stick with just using `_COPY_META_FIELDS` for now.

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64480755).

File: torch/_inductor/pattern_matcher.py [hunk 2]
```diff
@@ -253,6 +262,10 @@ def replace_by_example(
             replacement = trace_fn(
                 replacement_fn, torch.fx.map_arg(args, lambda arg: arg.meta["val"])  # type: ignore[arg-type]
             )
+            if len(self.nodes) > 0:
+                for n in replacement.graph.nodes:
+                    _transfer_meta(new_meta=n.meta, old_meta=self.nodes[0].meta)
```
Comment (aakhundov) [inline torch/_inductor/pattern_matcher.py:267]: What if we have multiple nodes in `self.nodes`? Can we end up with more confusion downstream?

Comment (davidberard98) [reply torch/_inductor/pattern_matcher.py:267]: Yes. I'd argue it's still better than having nothing though:
* In practice, it looks like very few replace_by_example patterns actually have multiple source nodes.
* The main case where you see this is in the (`add(z, mm(x, y)) -> addmm(z, x, y)` case.)
* original_aten may be confusing, but you'll also get other information like source location that will still be useful.

Do you have a suggestion on a better way to get this information

Comment (davidberard98) [reply torch/_inductor/pattern_matcher.py:267]: Or alternatively, how about we only preserve metadata if self has exactly 1 node?

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64480755).

File: torch/_inductor/pattern_matcher.py (hunk 2 above)
Comment (aakhundov) [reply torch/_inductor/pattern_matcher.py:267]: > Or alternatively, how about we only preserve metadata if self has exactly 1 node?

IMO, this is potentially less confusing.

Comment (davidberard98) [reply torch/_inductor/pattern_matcher.py:267]: I've changed the PR to do this (only preserve metadata if self has exactly 1 node)

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64480755).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64480755).

**Review by aakhundov — APPROVED** [review]
Accepting as the remaining test failures seem flaky and the internal CI looks good, too.

**Comment by yf225** [comment]
@davidberard98 I was watching CI this weekend - suspect this might be breaking `python test/inductor/test_pad_mm.py PadMMTest.test_original_aten_preserved_pad_mm` (https://ossci-raw-job-status.s3.amazonaws.com/log/31774004246)

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your PR but the new test_original_aten_preserved_pad_mm test runs OOM in trunk https://hud.pytorch.org/pytorch/pytorch/commit/fb44658415e50b5be6a187ff3f14243c0fdf3daf' -c nosignal

inductor/test_pad_mm.py::PadMMTest::test_original_aten_preserved_pad_mm [GH job link](https://github.com/pytorch/pytorch/actions/runs/11419127648/job/31773997247) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/fb44658415e50b5be6a187ff3f14243c0fdf3daf)

If possible, I think a smaller tensor is needed.  Otherwise, you could add test_pad_mm into https://github.com/pytorch/pytorch/blob/main/test/run_test.py#L225 to make it run serially and can use all the GPU memory.

**Comment by davidberard98** [comment]
Shrunk the test to use 2MB. Tests are passing and the memory is smaller, so I'll try merging again.

**Pull request closed without merging**